### ISSUE TITLE
Better interoperability with System.Numerics.Vectors

### ISSUE THREAD

**Issue by vpenades** [dotnet/Microsoft.Maui.Graphics#136]
Along with my previous issue #135 , There's some other types that could be mapped to System.Numerics.Vectors in some nice ways:

- Color: Is exactly the same as `System.Numerics.Vector4`
- PointF:  Is exactly the same as `System.Numerics.Vector2`
- SizeF: Is exactly the same as `System.Numerics.Vector2`
- RectangleF: could have constructors using two `System.Numerics.Vector2`

I thought Microsoft learned the lesson:  Since the dawn of graphics APIs, every time Microsoft rolled a new API, it has reimplemented again and again the same primitive types.

So now we're going to have:
PointF, SizeF, RectangleF and so on in System.Common.Drawing (GDI).
Point, Vector, Rectangle and so on in System.Windows (WPF).
PointF, SizeF, RectangleF and so on in Microsoft.MAUI.

So, this might come at a surprise, but people has been using System.Numerics  Vector2 and Matrix3x2 as the de-facto standard for Point, Size, and AffineTransforms. You just need to take a look at ImageSharp library to realize that.

So please, don't force everybody to have to do conversion extensions between System.Numerics and MAUI types. Please implement implicit converters, or, at the very least, constructors accepting Vectors.

**Comment by jonlipsky** [comment]
@vpenades Thanks for this feedback, and I think I/we can address this.  Don't get upset at Microsoft. If you want to get upset with anyone, get upset with me.  99% of this code was pulled from my 10+ year old code base that pre-dates System.Numerics being available in Xamarin, and switching over to System.Numerics has always been lower on my to-do list than adding features to my apps that used this library.

**Comment by vpenades** [comment]
@jonlipsky Given that Maui is going to become the groundwork for many years to come, I think it's important to settle on a good foundation. Little decissions now may affect the quality of the code for many years into the future.

I don't know how settled is the API right now, If I would be me, I would replace AffineTransform by Matrix2x3. If that's not possible, at the very least I would make AffineTransform a readonly struct, and fully exchangeable with Matrix2x3, not only by adding implicit converters, but also, allowing direct casting with Span<T>.

The same applies for all other types like Size, Point and so on.

Not sure how to add a dependency on System.Numerics for the current nuget package.... but if agreed, I could offer PRs for the rest of the stuff like implicits

**Comment by jonlipsky** [comment]
@vpenades I don't disagree with anything you said, and I think all of your proposed changes are great additions.  I would happily accept and collaborate with you on this PR.

**Comment by vpenades** [comment]
As usual... easier to be said than done.

Looking at the AffineTransform definition and methods, it has some methods that require it to be a class. In particular all the methods that modify the values of the class, basically all the Set* methods.

So either AffineTransform is ditched completely, or is kept and remains as a class. My convern with AffineTransorm being a class is that it requires a GC allocation for every instance, which can be noticeable in some scenarios.

**Comment by jonlipsky** [comment]
You're completely correct on the class vs struct issue with AffineTransform, and it all depends on how you use it.  For my use cases, having it as a class improves performance and having it mutable matched my needs.  (I tried it both ways before I settled on making it a class)

With this library being shared with a broader audience, I'm not sure that everyone will understand the nuance of how and why it's implemented the way it currently it.

My proposal would be to keep all of the set methods, but have them return a new AffineTransform/Matrix with that modification made.

**Comment by vpenades** [comment]
Performance of classes is better _as long as you don't trigger a GC collection. Ideally, drawing stacks should be as GC collection free as possible. I suspect with AffineTransform being a class you can end with thousands of instances waiting to be collected after some minutes.

I agree with you the Set methods to return the mutated value

**Comment by jonlipsky** [comment]
Exactly, if you know what you're doing, classes can be better.  Like I said earlier, all of this code came from an existing codebase completely written by me, tuned for my use cases.

**Comment by vpenades** [comment]
@jonlipsky is this proposal still being considered?  I was wondering if it could be possible to add before release.

looking at the code, the base library would require a dependency on System.Numeric.Vectors (at least for NetStandard2.0) , no sure if this is an issue...  and that's one of the reasons why I didn't try a pull request.... because I don't know if adding dependencies as a go/no go.

The idea would be to add System.Numeric.Vectors only for the target frameworks that needed it... most frameworks today have inbuilt support.

**Comment by jonlipsky** [comment]
@vpenades I would still be open to it, and Maui.Graphics already has a dependency on System.Numerics.Vector because of some Path flattening code I added earlier this summer.

**Comment by vpenades** [comment]
Would you also consider implicit converters for tuples?

consider this:

```c#
dx.DrawCircle( new Point(10,10), 10, Red);
```
with an implicit `(float,float)` to `Point` converter, you could also do:

```c#
dx.DrawCircle( (10,10), 10, Red);
```

Which removes a LOT of verbosity when writing drawing code.

**Comment by vpenades** [comment]
I've just commited the pull request with the system.numerics interoperability. Feel free to add/remove whatever you need.

Some stuff I've noticed while writing the pull request.

AffineTransform has two methods that are redundant: SetMatrix and SetTransform, I guess one of the two should be marked obsolete in favour of the other.

There's no need to explicitly reference System.Numerics.Vectors for all TargetFrameworks, since some of the frameworks have in built support. It is possible to conditionally reference the package only for those cases where it's required.

**Comment by jonlipsky** [comment]
@vpenades With regards to implicit conversions from tuples, my preferred way would be to reduce verbosity is to add additional extension methods, so that you could end up with something like this:

```c#
dx.DrawCircle( 10, 10, 10, Red);
```

Thanks for pointing out the redundant `SetMatrix` and `SetTransform` methods.  That's a result of some old dead in what Maui.Graphics is based on.
